format/Flight.proto

```diff
@@ -301,6 +301,36 @@ message Location {
  */
 message Ticket {
   bytes ticket = 1;
+  // Data representing some part of the data retrievable by the ticket.
+  //
+  // `inlined_completeness` indicates what part of the data retriavable
+  // by the ticket this represents. This is provided as an optimization for
+  // client/server applications that want to reduce latency to first result
+  // without requiring another RPC round-trip to retrieve the ticket.  applications
+  // built on top of Flight are responsible for any negotiation necessary on whether
+  // inlining data is appropriate.
+  //
+  // The size of inlined_data is expected to be small (typically less then 1MB) 
+  // and inlining too much data across tickets can run into underlying transport 
+  // limitations.  Furthermore, since the data is expected to be small, implementations 
+  // are less likely to optimize for zero-copy in these cases. 
+  //
+  // The inlined data is not expected to contain schema metadata. The schema
+  // should be identical to the schema provided on FlightInfo. 
+  repeated FlightData inlined_data = 2;
+  enum InlinedCompleteness {
```

Comment:
Since the Ticket is already an opaque data blob that is interpreted by the application, I am not sure how much extra value adding a generic `inlined_data_completeness` would be.

For applications that need to send `NO_INLINED_DATA` they simply omit `inlined_data`

For applications that need to send data, they could include `inlined_data` and could distinguish between "complete" and "sample data" somehow in the application defined `Ticket` payload anyways.

Reply:
Ticket is meant to be opaque to the client, so it may not be appropriate to use for this? (I've always thought of it as analogous to cookies or HTTP session state - the client carries it and passes it back to the server, but doesn't manipulate it.) That said I do agree that a general application metadata field somewhere in FlightInfo would likely be useful.

Reply:
Since the Ticket is already an opaque data blob that is interpreted by the application, I am not sure how much extra value adding a generic inlined_data_completeness would be.

> For applications that need to send NO_INLINED_DATA they simply omit inlined_data

I'm going to simplify this to make a nested message for inlined_data which cleans this up a little bit.

> For applications that need to send data, they could include inlined_data and could distinguish between "complete" and "sample data" somehow in the application defined Ticket payload anyways.

I agree with David on the opaqueness of Ticket which is my understanding also.  I think the benefit of directly modeling complete vs incomplete, is at least for incomplete data, we can completely elide the details from consumers.  Users could directly pass the ticket to DoGet and the framework can recognize there is inlined data present and its complete and just use that without existing clients making any code changes.  I haven't started implementing this but that was the approach I was going to look at first to see if it is feasible.

For sampled data it is always going to require a client side change to make use of it unless we the framework provides stronger guarantees (i.e. data is always the same exact data as the first part of the stream).

Reply:
I am missing the idea that the client can be general and not bound to a specific server implementation. I probably don't understand how that would work in practice, because if for example the client needs to call `DoGet` to start the server computing something, the client needs to provide a `Ticket` it got somewhere 🤔  

I see now that perhaps `FlightDescriptor::cmd` is perhaps the correct place

Reply:
I don't think a client can easily be fully general, but the flow is meant to go GetFlightInfo(FlightDescriptor) -> FlightInfo, then DoGet(Ticket), where the Ticket comes from the FlightInfo.

Reply:
> DoGet to start the server computing something, the client needs to provide a Ticket it got somewhere 🤔

This is a good point also that there is an assumption for this that most of the work starts as part of GetFlightInfo or before.  This type of thing would not directly apply if their  computation didn't start until DoGet. 

At least for SQL and other types of storage systems I think this assumption is reasonable but agree not universally applicable.

Reply:
I think I / we may have misunderstood the expected flow for using Flight -- hopefully as we migrate to using FlightSQL (as we plan to do for IOx) we'll end up conforming more closely to the intended use

Reply:
You may be interested in https://github.com/apache/arrow/pull/12636 where we try to clarify some of these things, also https://github.com/apache/arrow/pull/12616 which tries to clarify it in the context of Flight SQL